src/mpid/ch4/shm/xpmem/xpmem_control.c

```diff
@@ -74,34 +72,24 @@ int MPIDI_XPMEM_ctrl_send_lmt_rts_cb(MPIDI_SHM_ctrl_hdr_t * ctrl_hdr)
      * we increase its refcount at enqueue time. */
     root_comm = MPIDIG_context_id_to_comm(slmt_rts_hdr->context_id);
     if (root_comm) {
-        int continue_matching = 1;
-        while (continue_matching) {
-            anysource_partner = NULL;
-
+        while (TRUE) {
             rreq = MPIDIG_dequeue_posted(slmt_rts_hdr->src_rank, slmt_rts_hdr->tag,
                                          slmt_rts_hdr->context_id,
                                          &MPIDIG_COMM(root_comm, posted_list));
-
-            if (rreq && MPIDI_REQUEST_ANYSOURCE_PARTNER(rreq)) {
-                /* Try to cancel NM parter request */
-                anysource_partner = MPIDI_REQUEST_ANYSOURCE_PARTNER(rreq);
-                mpi_errno = MPIDI_anysource_matched(anysource_partner,
-                                                    MPIDI_SHM, &continue_matching);
-                MPIR_ERR_CHECK(mpi_errno);
-
-                if (continue_matching) {
-                    /* NM partner request has already been matched, we need to continue until
-                     * no matching rreq. This SHM rreq will be cancelled by NM. */
+#ifndef MPIDI_CH4_DIRECT_NETMOD
```

Comment:
I just realized that this code is unnecessary if we're in the XPMEM module. If `MPIDI_CH4_DIRECT_NETMOD` is true, this code isn't getting compiled anyway.

Reply:
You are correct!

We could make a patch for it, or -- since it doesn't have any effect here, we could wait for next refactor.

If we shift the rndz protocol to mpidig -- #4330 -- then we don't really need deal with message matching here alltogether.

Reply:
That is essentially how OFI does it. But having LMT inside RNDZ cost one extra round-trip of am_send_hdr. Eventually I am thinking we should have `OFI_rndz_send`, `UCX_rndz_send`, `XPMEM_rndz_send`, `POSIX_rndz_send`..., that integrate the two protocols. The common part -- RTS, CTS, and message matching part should be mpidig inline routines, and each specific rndz can call those routines plus setting up stuff in their own specific ways.

Reply:
Alternatively, instead of specific RNDZ protocols, we could have general RNDZ protocol at mpidig-layer but with pre-registered send-side callback and target-side callback. Then OFI for example, simple can call `MPIDIG_do_rndz_send(int rndz_id, ...)` where rndz_id is preregistered with ofi callbacks to setup rdma and do-rdma-read respectively.

Reply:
I don’t mean moving the code. I mean the `ifdef` has no effect. It will never be false. 😀 We can just delete that line.

Reply:
I understand. I was simply adding notes :)